### ISSUE TITLE
Dependency Support

### ISSUE THREAD

**Issue by joshughes** [saltstack/kitchen-salt#29]
Many of my formula depend on other formula to setup a service exactly how it will be in my infrastructure. For instance I may need an Apache Reverse Proxy in front of a Tomcat Service. I have my Apache formula set that up in my test environment. 

I wrote the following vagrant plugin to solve that issue for me. https://github.com/joshughes/vagrant-saltdeps

I would be nice to have similar functionality in kitchen-salt. I am happy to help do the work, but being unfamiliar with test-kitchen I am not sure where something like dependency support should be implemented. Chef uses berkshelf and maybe salt needs a stand alone project like that? 

If there is precedence for test-kitchen to manage dependencies, then I would rather do it in kitchen-salt than manage an entire other project just for managing dependencies.

**Comment by simonmcc** [comment]
There are currently 2 ways to handle dependent formula in kitchen-salt, neither of them well documented,  so I'd welcome some help in make them better documented & more easily found!

Method 1 - dependencies hash
https://github.com/simonmcc/kitchen-salt/blob/master/provisioner_options.md#dependencies
a list of hashes specifying dependencies formulas to be copied into the VM. e.g. [{ :path => 'deps/icinga-formula', :name => 'icinga' }]
Example: https://github.com/simonmcc/beaver-formula/blob/test-kitchen/.kitchen.yml#L56

Method 2 - vendor_path
https://github.com/simonmcc/kitchen-salt/blob/master/provisioner_options.md#vendor_path
path (absolute or relative) to a collection of formula reuired to be copied to the guest
Example: https://github.com/simonmcc/beaver-formula/blob/test-kitchen/.kitchen.yml#L60

The one thing that differs from vagrant-saltdeps is that I assume you've already got all the formula played out on disk on the host & that we just need to copy them down to the test-kitchen guest to make them useable. (that's largely because there is no unified tool in salt for doing that - we use an in house tool called gitshelf, https://github.com/gitshelf/gitshelf, which is very similar to vagrant-saltdeps in that it takes a yaml file & lays the repos out on disk)

If you were looking at using kitchen-salt inside a CI/CD pipeline, the job prep step could arrange all the formula for you from a gitshelf.yml and then run kitchen?

**Comment by joshughes** [comment]
@simonmcc thanks for all the great information. I do believe you are correct, I will have to experiment a bit but I don't see any reason why gitshelf wouldn't work.

Another problem I am faced with right now is having to maintain many Vagrantfiles with the same information. Does test kitchen support having what I would call a 'Base' configuration. For example I need to upgrade to the next version of salt. Being able to modify this in one place and have all my tests inherit from this single place would be great. 

Once again using action hooks in the Vagrant Plugin infrastructure was how I was thinking of implementing this in my vagrant-saltdeps plugin... and having it get and merge in a 'Base' Vagrantfile before vagrant brings up the virtual machine. 

Also as I go through this process I will make sure to take notes and contribute back to the documentation. I have a few colleges who are facing similar issues and would also contribute back to making a robust solution.

**Comment by thedrow** [comment]
What about using https://docs.saltstack.com/en/latest/topics/spm/index.html
It seems like it's the right way to implement this feature for newer salt versions.

**Comment by simonmcc** [comment]
Wow, yes that would probably be the best fit, I missed that being announced as I've stepped away from Salt and on to Ansible due to a change in technical direction at $work.

**Comment by thedrow** [comment]
@simonmcc Do you have time to maintain this project? If not, let's create an issue for a call to new maintainers and link to it from the README file.

**Comment by joshughes** [comment]
So I went ahead and implemented what I needed in https://github.com/joshughes/vagrant-saltdeps. The main thing we are trying to accomplish is CI for our salt formula. So while Salt Package Manager helps with dependencies between formula, you still need a consistent platform to test all your salt code. 

For that we use vagrant, and having a base Vagrantfile that we then merge for each project means we have a central place to manage things like salt-version, ami-id, and the box we are using. So I can update the salt version in a single place and let CI test all my salt formula to make sure I am good for an upgrade. 

I haven't worked with test kitchen before, but I would be happy to try to port some of these features over.

**Comment by thedrow** [comment]
@joshughes That would be awesome.

**Comment by simonmcc** [comment]
@thedrow I don't get as much time as I would like, so I'm open to handing it over to people who do, or at the very least increase the number of people with merge rights.  I went as far as creating https://github.com/kitchen-salt/kitchen-salt in preparation for this a while back, but got side tracked on getting some ci working (which now provides valuable feedback on incoming changes)

**Comment by epcim** [comment]
BTW: I am going to extend dependencies in kithen-salt. Considering to support existing + SPM, (apt,yum packages). 

We are now using a `medatada.yml` on formula repo level to declare dependencies. Pkgs has pkg dependencies, so less work.

If I am aware spm itself does not have place where to store it's dependencies. Am I right?

Example:
https://github.com/tcpcloud/salt-formula-git/blob/master/metadata.yml

**Comment by ek9** [comment]
Just giving my two cents:

An alternative to using gitshelf or `saltdeps` would be just to use [GitFs Remotes in Salt](https://docs.saltstack.com/en/latest/topics/tutorials/gitfs.html). This is supported by salt natively, you only need to make sure to have python / git dependencies setup.

Once that is done, you can configure git remotes:

```yaml
gitfs_remotes:
  - git://github.com/example/first.git
  - https://github.com/example/second.git
```

It supports various global and per-remote configuration options (i.e. you can point it to branches, tags or you can only make it use one specific directory from `git` repository). Additionally, if you use [salt-formula](https://github.com/saltstack-formulas/salt-formula) then it can do the rest of the configuration.

I've looked to `vagrant-saltdeps` and I appreciate the effort. It's great if it solves problem and works for someone, but I believe attaching the dependency management to vagrant directly provides a lot of additional hurdles to come accross.

**Comment by gtmanfred** [comment]
Looks like we can close this now that i have merged the following?

https://github.com/saltstack/kitchen-salt/pull/93
https://github.com/saltstack/kitchen-salt/pull/78

Thanks,
Daniel
